[EAGLE-4744] Added Hf loader and Handled Concepts in model upload

luv-bansal opened the pull request:
<!---
The PR description should include WHAT, WHY, HOW it does things and how this PR is tested.
-->

 ### What 
<!--
You should try to state the WHAT in PR title only.
Your PR title should describe in short WHAT this PR does, e.g. "[AA-1] Add new Clarifai cool feature".
Uncomment the `### What` section above, if you decide to add more details about WHAT this PR does.
For example, you may choose to describe the WHAT in more detail in the below cases.
* If the PR title is too short to include what your PR does;
* If your PR does more than one thing, the title may not have enough space to include everything. In this case, consider creating multiple PRs to separate work and make it easier for everybody to understand. Alternatively, mention a list of things what this PR does in this section.
-->
- Added HuggingFaceLoader in utils
- Added methods in model_upload for download_checkpoints and handling concepts for classification, detection models
- Changed `--model_path` flat in `model_upload.py` inplace of `-folder`

### Why
<!-- Why is this PR trying to accomplish said thing? This is useful to understand your intention why this code should be merged to master. -->
*

### How
<!-- How is this PR trying to accomplish said thing? This is useful to understand the details of the code. The code reviewer will check this section to assist them with code review. -->
*

### Tests
<!-- How is this PR tested? Write some tests to convince yourself and the reviewer that your code works. Sometimes, no tests are necessary, but we should always ask the question "Can I add some tests for this PR?". -->
*

### Notes
<!-- Add any additional notes here. -->
*

zeiler commented on clarifai/runners/dockerfile_template/Dockerfile.cuda.template:
```diff
@@ -149,14 +118,11 @@ WORKDIR /app
 COPY . /app/model_dir/${name}
 
 # Copy the checkpoints directory into the distroless image.
-COPY --from=build /app/checkpoints /app/model_dir/${name}/1/checkpoints
```
it wouldn't be from=build right?

luv-bansal replied:
oh right, just corrected it

luv-bansal replied:
I just realized that we don't need to add the COPY checkpoints line in the Dockerfile, since we're already using `COPY . /app/model_dir/${name}` a line before. That line will also copy the checkpoint directory into the final folder of the container. What do you think?

zeiler replied:
we might want to separate out the code from the checkpoint though in the future as copying in the code after the checkpoint would re-use most layers if you'd iterating on the same model code a few times hopefully it can re-use the checkpoint layer.

zeiler commented on clarifai/runners/models/model_upload.py:
```diff
@@ -11,9 +11,15 @@
 class ModelUploader:
   DEFAULT_PYTHON_VERSION = 3.11
+  CONCEPTS_REQUIRED_MODEL_TYPE = [
+      'visual-classifier', 'visual-detector', 'visual-segmenter', 'visual-embedder',
```
the embedders shouldn't need concepts.

zeiler commented on clarifai/runners/models/model_upload.py:
```diff
@@ -212,13 +273,15 @@ def monitor_model_build(self, model_version_id):
 if __name__ == "__main__":
   parser = argparse.ArgumentParser()
-  parser.add_argument('--folder', type=str, help='Folder to tar and upload', required=True)
+  parser.add_argument(
```
https://github.com/Clarifai/clarifai-python/tree/master/clarifai/runners/models i wonder if we should add a README.md to that folder to describe how to upload a model? Or should we just add that once you have the examples and we point to that repo in the README?

luv-bansal replied:
Yeah, I think it would be better once we have examples in the `clarifai/examples`

zeiler reviewed (approved):
just the dockerfile COPY from then you can merge.

luv-bansal commented:
Added some more changes:
- Added Clarifai dependency in Dockerfile, before it was missing 
- Removed COPY checkpoints line in the Dockerfile, since we're already using `COPY . /app/model_dir/${name} `a line before. That line will also copy the checkpoint directory into the final folder of the container
- Improve handling of importing `huggingface_hub` in HF loader
